Add Submissions Model and functionality

whyderrick opened the pull request:
This PR builds out the second core model for this app and the join model between it and the `Poem` model. It also does some refactoring of PR #1 to bring what I now think are better practices.

I'd particularly love feedback on which additional cases I should anticipate in the specs.

whyderrick committed:
Create `Submission` & `SubmissionPacket` models

This commit creates 2 new models, `Submission` and `SubmissionPacket`.
Submission is the second core model opposite poem in the app and the
`SubmissionPacket` is the join model between `Poem` and `Submission`.
`SubmissionPacket` contains a status property that can't be calculated,
so a `has_and_belongs_to_many` association wasn't feasible.

`Submission::STATUSES` is a list of potential submission states taken
directly from Submittable's site. Restricting to those types will ease
the integration with Submittable down the road.
 `Poem::STATUSES` are all states I think make sense for a poem lifecycle
so far.

Finally, this commit removes an option from the SimpleForm initialize
about invalid handling that proved redundant alongside the errors
partial that exists in the app.

*Lessons and takeaways*
- Use constants where they make sense instead of jumping to the step of
creating helper methods.
- When passing params through strong_params, if the param value is an
array, that must be indicated in the strong_params method
- SimpleForm `associations` will create associated objects based on ids
with no extra input

whyderrick commented on app/views/poems/_form.html.erb:
```diff
@@ -1,13 +1,8 @@
-    form.select :status,
-    ["Available", "Pending", "Accepted – Forthcoming", "Published"]
-  %>
+  <%= form.select :status, Poem::POEM_STATUSES %>
```
not sure if it's kosher to use model constants in the view instead of wrapping them in a helper. Writing that makes me think not because it seems more tightly coupled, but I'm just responding to my gut feeling on it.

JoelQ replied:
I'm 👍  on using model constants in the view

thorncp commented on app/models/poem.rb:
```diff
@@ -1,5 +1,8 @@
+  has_many :submissions, through: :submission_packets
 
   validates :title, presence: true
   validates :status, presence: true
```
Should this also include an inclusion validation that uses `POEM_STATUSES`?

whyderrick replied:
Yup! Thanks.

thorncp commented on app/models/submission.rb:
```diff
@@ -0,0 +1,18 @@
+  has_many :submission_packets, dependent: :destroy
+  has_many :poems, through: :submission_packets
+
+  validates_presence_of :title 
```
It looks like `validates :attribute, presence: true` is used elsewhere in the app. Thoughts on using that here for consistency?

gylaz commented on app/models/submission.rb:
```diff
@@ -0,0 +1,18 @@
+
+  validates_presence_of :title 
+  validates :status, presence: true, inclusion: { in: SUBMISSION_STATUSES }
+  validates_presence_of :submitted_to
```
`validates` is the new (and preferred) way of defining validations. Looks like you used it on line `16`, any reason you didn't do it here?

whyderrick replied:
Not realizing `validates` was preferred in all cases. I used it on 16 because I was assigning multiple validations to a single attribute, but used `validates_presence_of` in other places because it seemed more descriptive. Easy fix.

gylaz commented on app/views/home/index.html.erb:
```diff
@@ -14,5 +14,33 @@
+      </table>
+    <% end %>
+    <%= 
+      link_to t("submissions.actions.new"),
```
Most view code I've seen places this line together with the opening `<%=`.

whyderrick replied:
like this (1):
```
    <%= link_to t("submissions.actions.new"),
       new_submission_path,
       class: "btn button"
    %>
```
like (2)

```
    <%= link_to t("submissions.actions.new"),
            new_submission_path,
            class: "btn button"
    %>
```
or like (3)
```
    <%= link_to t("submissions.actions.new"),
              new_submission_path,
              class: "btn button"
    %>
```
Enabling Hound will help, but there are indentation norms that I still mix up. (1) looked weird to me with the path starting to the left of the `link_to` call, so I did this initially. I'd say 2 looks cleanest. (though GH added in spacing very different from what I provided in the raw)

JoelQ replied:
I typically write it as:

```erb
<%= link_to(
  t("submissions.actions.new"),
  new_submission_path,
  class: "btn button"
) %>
```

Not sure if that is the preferred way or not 😄

gylaz replied:
I actually like what @JoelQ proposes so +1. Most of Rails code I've seen uses:
```erb
<%= link_to t("submissions.actions.new"),
  new_submission_path,
  class: "btn button" %>
```

gylaz commented on app/views/submissions/_form.html.erb:
```diff
@@ -0,0 +1,12 @@
+<%= simple_form_for @submission  do |form| %>
+  <% if @submission.errors %>
```
I believe this will always yield `true`, as an AR object will always have `errors`. So, you might want a `present?` check.

JoelQ replied:
Agreed. Truthy checks like this can bite you in lots of subtle ways.

whyderrick replied:
👍

gylaz commented on app/views/submissions/_form.html.erb:
```diff
@@ -0,0 +1,12 @@
+  <%= form.input :submitted_to %>
+  <%= form.select :status, Submission::SUBMISSION_STATUSES %>
+  <%= form.association :poems, as: :check_boxes %>
+  <%= form.input :user_id, as: :hidden, input_html: { value: current_user.id } %>
```
Any reason for passing this in as opposed to setting it on the back-end?
This looks like a security issue to me, as someone can modify the hidden element and send a submission on behalf of another user.

whyderrick replied:
No good reason at all. Updated this to be done in the controller.

gylaz commented on app/views/submissions/show.html.erb:
```diff
@@ -0,0 +1,11 @@
+<div>
+  <h2><%=@submission.title%></h2>
+  <h3><%=@submission.submitted_to%></h3>
```
There are a few place like this, but there should be a whitespace after `=` and before closing `%`.

gylaz commented on app/views/submissions/show.html.erb:
```diff
@@ -0,0 +1,11 @@
+  <h3><%=@submission.submitted_to%></h3>
+  <h3>Poems in this submission</h3>
+  <ul>
+    <% if @submission&.poems&.each do |poem| %>
```
Thoughts on avoiding all the `&` by ensuring we always return something in `@submission` (even if its `Submission.new`)?

JoelQ replied:
I think `@submission` is guaranteed to be present here? In the controller you use `find_by` which will raise an error if it doesn't find the submission (ultimately resulting in a 404)

whyderrick replied:
Yup. I mistakenly thought that calling `each` on an empty array produced an error too.

tabfugnic replied:
@JoelQ I don't think `find_by` raises an error. Both `find` and `find_by!` raise errors, but `find_by` will return `nil`

JoelQ replied:
oops! You are correct @tabfugnic. The controller actually uses `find` (as it should) not `find_by` and will raise if there is no submission.

mike-burns replied:
> I mistakenly thought that calling each on an empty array produced an error too.

To really drive the point home, I recommend the 1985 classic [How to Replace Failure by a List of Successes](https://rkrishnan.org/files/wadler-1985.pdf), by Wadler.

gylaz commented on db/migrate/20170421215941_create_submission_packets.rb:
```diff
@@ -0,0 +1,9 @@
+    create_table :submission_packets do |t|
+      t.references :poem, foreign_key: true, index: true, null: false
+      t.references :submission, foreign_key: true, index: true, null: false
+      t.string :status
```
Does it make sense to have no status set?

whyderrick replied:
I'm not sure at this point. The status of an individual `SubmissionPacket` only changes when the parent `Poem` or `Submission` does, but there's no meaningful initial state. I'll make a placeholder initial status to avoid the having an empty column, though.

JoelQ replied:
If the status depends on values set on the parent, perhaps it should be calculated rather than stored in the database?

e.g.

```ruby
class SubmissionPacket < ActiveRecord::Base
  def status
    # calculate based on parent status
  end
end
```

This prevents you from having corrupt data such as having a status on parent change but then the corresponding change doesn't happen on the submission packet.

whyderrick replied:
That's an interesting idea, but I think if anything can be calculated. I don't think I can calculate the `SubmissionPacket#status` but I might be able to nix `Poem#status` and calculate that instead. 


I'll add a card to figure that out and table that for another PR. For now, I'll just update these to default to "Submitted".

Thanks to you and @gylaz  for the push here. It's helping me clarify the recs.

gylaz commented on db/schema.rb:
```diff
@@ -37,16 +37,35 @@
     t.string   "confirmation_token", limit: 128
     t.string   "remember_token",     limit: 128, null: false
-    t.index ["email"], name: "index_users_on_email", using: :btree
+    t.index ["email"], name: "index_users_on_email", unique: true, using: :btree
```
Where did this change come from?

whyderrick replied:
For some reason, last week when you suggested this change on PR #1 I made it directly on the initial migration instead of generating a new one. I thought it was a defensible idea at the time (probably because it was the first migration that would run and I knew no one had setup locally), but I'm not willing to defend it now and won't do it again.

gylaz replied:
It's okay to modify a migration that you're adding in a new branch. You just have to make sure to run/re-run it, so it produces the correct `db/schema.rb`. At this point, I would pull this out into a separate quick PR, explaining that this change was missed in a previous commit/PR.

whyderrick replied:
Done. #3 

Benefit: I learned that I could combine `add --patch` with a version of a file from a specific past commit without having to Google it.

gylaz commented on spec/features/submission_spec.rb:
```diff
@@ -0,0 +1,63 @@
+    fill_in :submission_submitted_to, with: submitted_to
+    fill_in :submission_status, with: submission_status
+    check poem.title
+  end
```
Where are these two methods used?

whyderrick replied:
I wrote them as helpers prematurely removed them from the spec to write code more specific to the tests and forgot to handle clean up.

gylaz commented on spec/features/submission_spec.rb:
```diff
@@ -0,0 +1,63 @@
+    check poem.title
+  end
+end
+
```
✂️

whyderrick replied:
👍

gylaz commented on spec/models/submission_packets_spec.rb:
```diff
@@ -0,0 +1,8 @@
+require 'rails_helper'
```
Single quotes here, but double on line 4. Any reason Hound isn't enabled on this repo?

JoelQ replied:
👍 to enabling Hound on the repo

whyderrick replied:
Thanks. It escaped me that I _could_ enable Hound here.

thorncp commented on app/views/submissions/_form.html.erb:
```diff
@@ -0,0 +1,12 @@
+  <%= form.input :submitted_to %>
+  <%= form.select :status, Submission::SUBMISSION_STATUSES %>
+  <%= form.association :poems, as: :check_boxes %>
+  <%= form.input :user_id, as: :hidden, input_html: { value: current_user.id } %>
```
I tend to put these kinds of variables into the controller itself, so a user can't mess with the html and assign it to something they shouldn't. Because `user_id` is a permitted param, they could edit it to be any user in the system.

A pattern I've done a few times now that I like is to have a different controller method for the create action, that injects the current user specific things. Something like

```ruby
def new_submission_params
  submission_params.merge(user_id: current_user.id)
end

def submission_params
  params.require(:submission).permit(:title, :submitted_to, :status, poem_ids: [])
end
```

Where the create action would use `new_submission_params` and the update action would use `submission_params`.

JoelQ replied:
👍 for doing this in the controller

whyderrick replied:
Thanks for the code sample here. I updated this in another controller differently, and I think this is more explicit.

thorncp commented on app/views/submissions/new.html.erb:
```diff
@@ -0,0 +1,2 @@
+<h2><%=t('submissions.actions.new')%></h2>
```
There's a neat shortcut that Rails provides if you name the locale keys after the view path. In this case we have `submissions/new.html.erb`, so if you do this in `en.yml`:

```yml
submissions:
  new:
    title: Create a new submission
```

the view can then use it like `<%= t(".title") %>` (the leading `.` is important)

whyderrick replied:
I knew that but hadn't connected it as the reason for the convention around key naming in the yml files. Thanks!

thorncp commented on app/views/submissions/show.html.erb:
```diff
@@ -0,0 +1,11 @@
+  <ul>
+    <% if @submission&.poems&.each do |poem| %>
+      <li><%=poem.title%></li>
+      <% end %>
```
This `end` seems extra. What does it close?

mike-burns replied:
Closes the `if`. The indentation on the line above is off.

How about adding Hound to this repo?

thorncp replied:
What about the `end` on the next line?

whyderrick replied:
This `end` closes the `each` the next `end` closes the `if`. But the conversation around it means it's certainly not clear or idiomatic.

thorncp replied:
Ah, I see it now. I think there was another comment somewhere similar to this, but we're in the show action so we can assume `@submission` exists here, and because `poems` is a `has_many` association, it shouldn't be `nil` either. I think this could simplified to

```
<% @submission.poems.each do |poem| %>
  <li><%=poem.title%></li>
<% end %>
```

thorncp reviewed (commented):
I don't have time to make it all the way through right now, but may be able to come back to it in a bit.

JoelQ commented on app/controllers/submissions_controller.rb:
```diff
@@ -0,0 +1,21 @@
+  end
+ 
+  def show
+    @submission = Submission.includes(:poems).find(params[:id])
```
Is `includes` necessary when you are only looking up 1 submission?

whyderrick replied:
No. It isn't. I think I used it primarily for the sake of using it after learning about it. There's no real need for an eager load, but I figured that while the view does require the associated poems, there was no harm in it.

JoelQ commented on app/models/poem.rb:
```diff
@@ -1,5 +1,8 @@
 class Poem < ApplicationRecord
+  POEM_STATUSES = ["Rough Draft", "Ready", "Submitted", "Accepted", "Published"]
```
You might want to freeze this array so no one else can mutate it (that happens more than you'd think 😖 )

mike-burns replied:
Not just the array, but also the string. I sometimes see `.map(&:freeze).freeze`.

thorncp replied:
Alternatively, add `# frozen_string_literal: true` to the top of the file.

JoelQ replied:
You'll still want to freeze the array even if you have `# frozen_string_literal: true`

JoelQ commented on app/views/home/index.html.erb:
```diff
@@ -14,5 +14,33 @@
       <p>Looks like you don't have any poems yet.</p>
     <% end %>
     <%= link_to t("poems.actions.new"), new_poem_path, class: "btn button" %>
+    <h2>Your Submissions</h2>
```
I see you use I18n elsewhere on this page. Should this string be localized too?

whyderrick replied:
Ideally, yup. Once I made sure I had a working translation file I moved onto more of the functionality in this PR. That's definitely something I'll be more mindful of going forward.

JoelQ commented on app/views/poems/_form.html.erb:
```diff
@@ -1,13 +1,8 @@
-    form.select :status,
-    ["Available", "Pending", "Accepted – Forthcoming", "Published"]
-  %>
+  <%= form.select :status, Poem::POEM_STATUSES %>
```
Should this be:

```ruby
form.input :status, collection: Poem::POEM_STATUSES
````

JoelQ commented on app/views/submissions/_form.html.erb:
```diff
@@ -0,0 +1,12 @@
+  <%= form.select :status, Submission::SUBMISSION_STATUSES %>
+  <%= form.association :poems, as: :check_boxes %>
+  <%= form.input :user_id, as: :hidden, input_html: { value: current_user.id } %>
+  <%= form.button :submit %>
```
I believe simple form gives you `form.submit` ?

JoelQ commented on app/views/submissions/_form.html.erb:
```diff
@@ -0,0 +1,12 @@
+<%= simple_form_for @submission  do |form| %>
```
As a general rule of thumb, it's good practice to avoid using instance variables in partials. That couples them to being rendered from views whose controllers defined the instance variable.

Instead, use local variables. That way the partial can be rendered from anywhere as long as you pass the correct locals.

whyderrick replied:
👍

JoelQ commented on db/migrate/20170421205530_create_submissions.rb:
```diff
@@ -0,0 +1,10 @@
+      t.string :title, null: false
+      t.string :status, null: false
+      t.string :submitted_to, null: false
+      t.references :user, foreign_key: true, index: true, null: false
```
❤️  all the `null: false`!

JoelQ commented:
This is a pretty long PR (`+249 −19`). It looks like it adds 3 separate new features:

1. List all submissions on the home page
2. A page with the details of a submission
3. The ability to create new submissions

In general, it is usually better to break up PRs into several smaller PRs when possible. This reduced scope makes the code:

* easier to write
* easier to review
* faster to merge

Although there is some overhead to working with lots of smaller branches, in my experience this approach increases my velocity versus one big "does it all" PR.

mike-burns commented on config/initializers/simple_form.rb:
```diff
@@ -46,7 +46,8 @@
     b.use :hint,  wrap_with: { tag: :span, class: :hint }
-    b.use :error, wrap_with: { tag: :span, class: :error }
+    # Commented out until styling kicks in
+    # b.use :error, wrap_with: { tag: :span, class: :error }
```
It's in git -- feel free to delete from the code and recover from git later.

whyderrick replied:
Yup. Perfectly sensible and pithy enough to make it clear why code shouldn't be commented out in any case.

mike-burns commented on spec/features/submission_spec.rb:
```diff
@@ -0,0 +1,63 @@
+      fill_in :submission_title, with: submission_title
+      fill_in :submission_submitted_to, with: submitted_to
+      select submission_status, from: :submission_status
+      submit_form
```
You might be interested in [the `formulaic` gem](https://github.com/thoughtbot/formulaic).

tabfugnic commented on app/views/submissions/_form.html.erb:
```diff
@@ -0,0 +1,12 @@
+
+  <%= form.input :title %>
+  <%= form.input :submitted_to %>
+  <%= form.select :status, Submission::SUBMISSION_STATUSES %>
```
Should this be taking advantage of simple_form? Something like 
```ruby
<%= form.input :status, collection: Submission::SUBMISSION_STATUSES %>
```

whyderrick replied:
:+1:

tabfugnic commented on spec/factories.rb:
```diff
@@ -3,4 +3,10 @@
+
+  factory :poem do
+    user 
+    title "The Best Poem"
```
I'd love to read it! :smile:

whyderrick commented:
@JoelQ Thanks for the feedback on the PR length. Any tips (preferably existing blog posts or the like) about how to appropriately break down PR when you're starting from a high-level spec? Or how to create a better starting feature spec than can pass before submitting the PR.

whyderrick committed:
Security and Convention Fixes

This commit makes a few changes:
- Updates the forms and controllers so that we're no longer passing a
`user_id` through the form as a hidden field and are instead assigning
the user in the controller.
- Updates the model validations to use `validates <field>, presence:
true` instead of `validates_presence_of`, because the latter is old
style and not preferred.
- Adds a new home.en.yml file for translations on the homepage

* Lessons*
- Passing the `user_id` of the current_user as a hidden form input is a
security risk. Don't do that anymore. Instead, the user should be
assigned in the controller. There are two strategies for it in this
commit, and the Hash merge is probably the preferable one.
- Use `validates` with options instead of validation specific methods
(unless using a custom validation)
- When creating constants, be sure to freeze string literals and arrays
that contain them whenever applicable

JoelQ commented:
@whyderrick looking at your spec, it doesn't require a list of submission on the home page. It also doesn't need the "show" page if you remove those last two expectations.

```ruby
expect(page).to have_flash_message(
  :notice,
  text: "Submission was successfully created."
)
expect(page).to have_text(submission_title)
expect(page).to have_text(poem.title)
```

This behavior could be built up iteratively.

1. PR 1 - User clicks link on home page, fills out form, and is redirected back to home page with a flash message (_same spec you have right now but without the last two expectations_).
2. PR 2 - User is now redirected to a show page when successfully creating a submission (_just add the two expectations back to the spec_)
3. PR 3 - User can view a list of submission on their profile and clicking on one leads to a detail page. (_this is unrelated to the creation flow. I would probably create a separate feature spec for this_)

Generally speaking, you can do the following to limit scope:

* Don't add extra features to the app not needed to pass your feature spec
* Make your feature spec have a smaller scope. This can be done via:
  * Removing/changing some of the expectations
  * Writing a feature spec that only goes through part of the flow (e.g only step 1/3 in a wizard flow). You can add the next steps in future PRs.

Unfortunately, I'm not aware of good blog posts on the topic (any ideas here @tabfugnic @gylaz ?). Derek Prior was talking about writing one for our blog the other day.

whyderrick committed:
Standardize forms and namespacing

This commit updates form partials to take locals instead of instance
variables, only renders the "errors" partial if object has errors
(instead of rendering it with conditional context), and uses the
SimpleForm convention for generating select inputs (instead of
`form.select`).

It also changes the name of model constants to rely on existing
namespacing instead of prefixing the constant name with its home model.

Finally, it adds timestamps to the models that have been created through
now. Sometimes we all forget things, and adds a default value to
`SubmissionPacket#status`.

* Lessons *
- Name variables in a way that's so descriptive you'll have no clue
understanding exactly what you meant when you look at this again in 6
months with absolutely no context.
- Partials shouldn't use instance variables. Pass locals instead.
- When splitting method call across lines, try to break them up in a way
that accentuates what's important. Even though it's more lines of code,
each permitted param gets its own line in `SubmissionController`.
- Don't be overzealous with prefixes. Namespacing exists to handle
chunks of ambiguity around what something like `STATUS` could refer to.
Prefixing everything breaks conventions without gaining much.
- Content inside erb snowcones gets spaces around it. Always.
- With SimpleForm, render a `select` with `form.input, collection:
<collection>` instead of `form.select`

Naming for descriptiveness and namespaces

- Josh pointed out that the `new_submission_params` method name wasn't
descriptive in a way that would help me when I'm tracing this code in a
few weeks/months. It's updated to something that makes its function
clearer.
- When splitting method call across lines, try to break them up in a way
that accentuates what's important. Even though it's more lines of code,
each permitted param gets its own line in `SubmissionController`.
- Don't be overzealous with prefixes. Namespacing exists to handle
chunks of ambiguity around what something like `STATUS` could refer to.
Prefixing everything breaks conventions without gaining much.

gylaz commented on app/controllers/poems_controller.rb:
```diff
@@ -4,17 +4,17 @@ def new
 
   def show
-    @poem = Poem.find(params[:id])
+    @poem = Poem.find(params[:id]) || Poem.new
```
`find` will raise an error if nothing is found so the `|| Poem.new` will never trigger.

jcmorrow replied:
Also, it's probably confusing to a user who attempts to view an existing poem and instead sees a new poem (presumably with a bunch of blank fields).

gylaz commented on app/controllers/submissions_controller.rb:
```diff
@@ -0,0 +1,27 @@
+
+  def show
+    @submission = Submission.includes(:poems).find(params[:id]) ||
+      Submission.new
```
Similar here.

gylaz commented on db/migrate/20170421205530_create_submissions.rb:
```diff
@@ -0,0 +1,10 @@
+class CreateSubmissions < ActiveRecord::Migration[5.0]
+  def change
+    create_table :submissions do |t|
```
Looks like this table is missing timestamps.

gylaz commented on db/migrate/20170421215941_create_submission_packets.rb:
```diff
@@ -0,0 +1,9 @@
+class CreateSubmissionPackets < ActiveRecord::Migration[5.0]
+  def change
+    create_table :submission_packets do |t|
```
timestamps?

jcmorrow commented on app/controllers/poems_controller.rb:
```diff
@@ -4,17 +4,17 @@ def new
   private
   def poem_params
-    params.require(:poem).permit(:title, :status, :user_id)
+    params.require(:poem).permit(:title, :status)
```
Nice change.

jcmorrow commented on app/controllers/submissions_controller.rb:
```diff
@@ -0,0 +1,27 @@
+
+  def submission_params
+    params.require(:submission).
+      permit(:title, :submitted_to, :status, poem_ids: [])
```
this is nitpicky but personally I would format this as

```
params.require(:submission).permit(
  :poem_ids: [],
  :status,
  :submitted_to,
  :title,
)
```

It makes the most interesting part (the params this form accepts) more clear.

whyderrick replied:
Nitpicking trains nuance, so I'm glad for it. 
Also, this is clearly better.

jcmorrow commented on app/controllers/submissions_controller.rb:
```diff
@@ -0,0 +1,27 @@
+  end
+
+  def new_submission_params
+    submission_params.merge(user_id: current_user.id)
```
This method isn't named very descriptively. I would either name it `user_submission_params` or just append this `merge` onto the original `submission_params` method.

whyderrick replied:
:+1:

jcmorrow commented on app/models/poem.rb:
```diff
@@ -1,6 +1,18 @@
+# frozen_string_literal: true
+
 class Poem < ApplicationRecord
+  POEM_STATUSES = [
```
Might be better to call this `STATUSES`? It will always be prefixed by the `Poem` namespace outside of this file so it will read well.

whyderrick replied:
Given the namespacing, I'm not seeing a counterargument. So yup. Thanks!

jcmorrow commented on app/models/submission.rb:
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+class Submission < ApplicationRecord
+  SUBMISSION_STATUSES = [
```
I would make the same argument here.

jcmorrow commented on app/views/home/index.html.erb:
```diff
@@ -11,8 +11,35 @@
       </dl>
     <% else %>
-      <p>Looks like you don't have any poems yet.</p>
+      <p><%= t(".no_poems") %></p>
```
👍

jcmorrow reviewed (approved):
Mostly nitpicky stuff, this is headed in a good direction 🏅

whyderrick commented:
![](https://github.trello.services/images/mini-trello-icon.png) [As a user, I can add submissions that I can track](https://trello.com/c/fOHOomLl/12-as-a-user-i-can-add-submissions-that-i-can-track)
